src/chain/unzip.rs

```diff
@@ -29,174 +29,82 @@ pub trait Unzippable: Sized {
-            .take_input::<Self>()?;
-
-        let targets = world.get::<ForkTargetStorage>(source).or_broken()?;
-        let target = (targets.0)[0];
-        if let Some(mut t_mut) = world.get_entity_mut(target) {
-            t_mut.give_input(session, inputs.0, roster)?;
-        }
-        Ok(())
-    }
-
-    type Prepended<T> = (T, A);
-    fn prepend<T>(self, value: T) -> Self::Prepended<T> {
-        (value, self.0)
-    }
-}
-
-impl<A: 'static + Send + Sync, B: 'static + Send + Sync> Unzippable for (A, B) {
-    type Unzipped = (Output<A>, Output<B>);
-    fn unzip_output(output: Output<Self>, builder: &mut Builder) -> Self::Unzipped {
-        assert_eq!(output.scope(), builder.scope());
-        let targets = Self::make_targets(builder.commands);
-
-        let result = (
-            Output::new(builder.scope, targets[0]),
-            Output::new(builder.scope, targets[1]),
-        );
-
-        builder.commands.add(AddOperation::new(
-            Some(output.scope()),
-            output.id(),
-            ForkUnzip::<Self>::new(ForkTargetStorage(targets)),
-        ));
-        result
-    }
-
-    fn make_targets(commands: &mut Commands) -> SmallVec<[Entity; 8]> {
-        SmallVec::from_iter([
-            commands.spawn(UnusedTarget).id(),
-            commands.spawn(UnusedTarget).id(),
-        ])
-    }
-
-    fn distribute_values(
-        OperationRequest { source, world, roster }: OperationRequest,
-    ) -> OperationResult {
-        let Input { session, data: inputs } = world
-            .get_entity_mut(source).or_broken()?
-            .take_input::<Self>()?;
-
-        let targets = world.get::<ForkTargetStorage>(source).or_broken()?;
-        let target_0 = *targets.0.get(0).or_broken()?;
-        let target_1 = *targets.0.get(1).or_broken()?;
-
-        if let Some(mut t_mut) = world.get_entity_mut(target_0) {
-            t_mut.give_input(session, inputs.0, roster)?;
-        }
-
-        if let Some(mut t_mut) = world.get_entity_mut(target_1) {
-            t_mut.give_input(session, inputs.1, roster)?;
+macro_rules! impl_unzippable_for_tuple {
+    ($($T:ident),*) => {
+        #[allow(non_snake_case)]
+        impl<$($T: 'static + Send + Sync),*> Unzippable for ($($T,)*)
+        {
+            type Unzipped = ($(Output<$T>,)*);
+            fn unzip_output(output: Output<Self>, builder: &mut Builder) -> Self::Unzipped {
+                assert_eq!(output.scope(), builder.scope());
+                let mut targets = SmallVec::new();
+                let result =
+                    (
+                        $(
+                            {
+                                // Variable is only used to make sure this cycle is repeated once
+                                // for each instance of the $T type, but the type itself is not
+                                // used.
+                                #[allow(unused)]
+                                let $T = std::marker::PhantomData::<$T>;
```

Comment:
This is a "hack" I had to create to allow iterating over all the types passed as an input without actually using them